README.rst
cosmo
=====

A cosmological calculator application using Electron and python.

Installation
------------
-   (Current) Successful installation instructions:
    -   Use an environment where *python2* is the default.
    -   `$ npm install`
    -   `$ npm rebuild zerorpc --runtime=electron --target=1.8.1`   # maybe not needed?
    -   `$ npm rebuild zmq --runtime=electron --target=1.8.1`       # maybe not needed?
    -   `$ ./node_modules/.bin/electron-rebuild`
-   When packaging (i.e. with `npm run dist`), *python2* must still be the default python!


- Run ``npm install``
  - This won't work with python versions > 3.0.0, so you may need to point to a particular python version, e.g.

    ``npm install --python=python2.7``

  or    

    ``npm install --python=/usr/bin/python``

  - The build of `zeromq` might also have a conflict between Node and Electron (or something?), and may require rebuilding, so this has been added to the npm 'postinstall':
    `npm rebuild zeromq --runtime=electron --target=1.7.2`
    where `1.7.2` is the particular version of electron being used.


CHANGES.rst
CHANGES
=======

Future / To-Do
--------------


Current
-------


[0.2.0] - 2017/12/28
--------------------
    -   Cosmological values now include: redshift, scale-factor, distance (luminosity & comoving), time (universe-age & lookback).
        -   This uses the `cosmopy` python package as a backend.
    -   Each parameter can be entered to retrieve all of the other parameters.
    -   Plot of cosmological parameters vs. redshift has been added.


[0.1.0] - 2017/12/23
--------------------
    -   The program uses Electron for the frontend with a minimal python backend to run the cosmological calculation.
    -   Currently a redshift can be entered and a luminosity distance will be calculated and returned in response.


app/face/plot.js
/*
 *
 *
 *
 */

/*    ========    SETTINGS    =========    */
const sprintf = require('sprintf-js').sprintf
const fs = require('fs')
const path = require('path');

var svg = d3.select('#simContainer');

var data, data_z, data_dl, data_len;
var path_dc, path_dl, path_tl, path_ta;
var line_dc, line_dl, line_tl, line_ta;
var focus;

var sleep_count = 0;
var retval = false;
var DATA_FILE_PATH;
const SLEEP_COUNT_MAX = 20;

var style = getComputedStyle(document.body);
var color_dc = style.getPropertyValue('--color-dc');
var color_dl = style.getPropertyValue('--color-dl');
var font_size = style.getPropertyValue('--font-size-small');

// These are strings
var width = svg.style("width");
var height = svg.style("height");

width = parseFloat(width.replace("px", ""));
height = parseFloat(height.replace("px", ""));
font_size = parseFloat(font_size.replace("px", ""));

var A_TICKS = [0.99, 0.9, 0.8, 0.5, 0.2, 0.05, 0.01];

var margin = {top: 40, right: 84, bottom: 50, left: 84};
margin.width = margin.left + margin.right;
margin.height = margin.top + margin.bottom;

const Z_RANGE = [0.01, 100.0];   // Range of redshift values
const D_RANGE = [1.0e+1, 1.0e5];  // Range of distance [Mpc] values
const T_RANGE = [1.0e-1, 20.0];  // Range of time [Gyr] values

const NZ_GRID = 3;
const ND_GRID = 3;
const NT_GRID = 2;

const APP_DIR = __dirname;

/*    ========    INITIALIZE OBJECTS    =========    */

// Storage for definitions, 'defs' tells SVG they are resources
var defs = svg.append("defs");

/*    ========    FUNCTIONS    =========    */

function format(num) {
    return num.toExponential(2);
}

// == Plot / Figure Initialization == //

function initPlot() {

    // Construct mask/overlay to monitor mouse over plot; and clip lines outside plot areas
    var mask = defs.append("clipPath")
        .attr("id", "mask")
        .style("pointer-events", "none")
        .append("rect")
        .attr('x', margin.left)
        .attr('y', margin.top)
        .attr('width', width - margin.width)
        .attr('height', height - margin.height)

    // Create a `g` called `plots` which we be clipped outside of the `mask`
    //    This is where all plotted lines should go, using:
    //    ``svg.select("#plots").append("path") ....``
    svg.append("g")
        .attr("id", "plots")
        .attr("clip-path", "url(#mask)");

    // == Scales and Axes == //
    axis_z = d3.axisBottom();
    axis_d = d3.axisLeft().ticks(5);
    axis_t = d3.axisRight().ticks(5);
    axis_a = d3.axisTop();

    scale_z = d3.scaleLog();
    scale_d = d3.scaleLog();
    scale_t = d3.scaleLog();

    scale_z = scale_z.domain(Z_RANGE)
        .range([0.0, width - margin.width]);
    scale_d = scale_d.domain(D_RANGE)
        .range([height - margin.height, 0]);
    scale_t = scale_t.domain(T_RANGE)
        .range([height - margin.height, 0]);

    axis_z.scale(scale_z);
    axis_a.scale(scale_z);
    axis_d.scale(scale_d);
    axis_t.scale(scale_t);

    // Convert to redshift values
    var a_ticks_z = [];
    for (var ii = 0, len = A_TICKS.length; ii < len; ii++) {
        a_ticks_z.push((1.0/A_TICKS[ii]) - 1.0);
    }
    // Set tick locations in redshift-scale, but label according to scale-factor
    axis_a.tickValues(a_ticks_z)
        .tickFormat(function(d) {
            return sprintf('%.2f', 1.0/(d + 1.0));
        });

    // d axis
    var _ax_d = svg.append("g")
        .attr("class", "axis")
        .attr("id", "axis_d")
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        .call(axis_d);

    // t axis
    var _ax_t = svg.append("g")
        .attr("class", "axis")
        .attr("id", "axis_t")
        .attr("transform", "translate(" + (width - margin.right) + "," + margin.top + ")")
        .call(axis_t);

    // z axis
    var _ax_z = svg.append("g")
        .attr("class", "axis")
        .attr("id", "axis_z")
        .attr("transform", "translate(" + margin.left + "," + (height - margin.bottom) + ")")
        .call(axis_z);

    // a axis
    var _ax_a = svg.append("g")
        .attr("class", "axis")
        .attr("id", "axis_z")
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        .call(axis_a);

    axisLabelTens(_ax_z);
    axisLabelTens(_ax_t);
    axisLabelTens(_ax_d);

    //  ==  Add Grid Lines  ==  //
    initGridLines();

    //  ====  Add Axis Labels  ====  //
    labelAxes();

    //  ====    Add Mouse-Interaction with Plot    ====  //
    initCrossHairs();

}

function initCrossHairs() {
    focus = svg.select("#plots").append('g')

    focus.append('line')
        .attr('class', 'focusLine')
        .attr('id', 'focusLine_z');
    focus.append('line')
        .attr('class', 'focusLine')
        .attr('id', 'focusLine_dl');
    focus.append('line')
        .attr('class', 'focusLine')
        .attr('id', 'focusLine_dc');
    focus.append('line')
        .attr('class', 'focusLine')
        .attr('id', 'focusLine_tl');
    focus.append('line')
        .attr('class', 'focusLine')
        .attr('id', 'focusLine_ta');

    svg.append('rect')
        .attr('class', 'overlay')
        .attr('x', margin.left)
        .attr('y', margin.top)
        .attr('width', width - margin.width)
        .attr('height', height - margin.height)
        .on('click', function() {
            var mouse = d3.mouse(this);
            var zz = scale_z.invert(mouse[0] - margin.left);
            calcAndUpdate(zz);
        });
}

function initGridLines() {
    // add z gridlines
    svg.select("#plots").append("g")
        .attr("class", "grid")
        .attr("id", "z")
        .attr("transform", "translate(" + margin.left + "," + (height - margin.bottom) + ")")
        .call(d3.axisBottom(scale_z)
            .ticks(NZ_GRID)
            .tickSize(-height + margin.height)
            .tickFormat("")
        );

    // add d gridlines
    svg.select("#plots").append("g")
        .attr("class", "grid")
        .attr("id", "d")
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        .call(d3.axisLeft(scale_d)
            .ticks(ND_GRID)
            .tickSize(-width + margin.width)
            .tickFormat("")
        );

    // add t gridlines
    svg.select("#plots").append("g")
        .attr("class", "grid")
        .attr("id", "t")
        .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        .call(d3.axisRight(scale_t)
            .ticks(NT_GRID)
            .tickSize(width - margin.width)
            .tickFormat("")
        );

}

function labelAxes() {
    // == Redshift Axis
    svg.append("text")
        .attr("id", "axis_z")
        .attr("transform",
            "translate(" + (width/2) + ", " + (height - margin.bottom + 40) + ")")
        .style("text-anchor", "middle")
        .style("font-size", font_size*1.2)
        .text("Redshift");

    // == Scalefactor Axis
    svg.append("text")
        .attr("id", "axis_z")
        .attr("transform",
            "translate(" + (width/2) + ", " + 14 + ")")
        .style("text-anchor", "middle")
        .style("font-size", font_size*1.2)
        .text("Scale Factor");

    // == Distance Axis
    svg.append("text")
        .attr("id", "axis_d")
        .attr("transform", "translate(15, " + (height/2) + ")" + "rotate(-90)")
        // .attr("dy", "1em")
        .style("text-anchor", "middle")
        .style("font-size", font_size*1.2)
        .text("Distance [Mpc]");

    // Comoving
    var temp = svg.append("text")
        .attr("id", "dc")
        .attr("transform", "translate(15, " + (height/2) + ")" + "rotate(-90)")
        .style("text-anchor", "middle")
        .text("Comoving");

    // Shift up
    temp.attr("dx", - temp.node().getBBox().width/2 - 10)
        .attr("dy", temp.node().getBBox().height + 4)

    // Luminosity
    var temp = svg.append("text")
        .attr("id", "dl")
        .attr("transform", "translate(15, " + (height/2) + ")" + "rotate(-90)")
        .style("text-anchor", "middle")
        .text("Luminosity");

    // Shift down
    temp.attr("dx", temp.node().getBBox().width/2 + 10)
        .attr("dy", temp.node().getBBox().height + 4)

    // == Time Axis
    svg.append("text")
        .attr("id", "axis_t")
        .attr("transform", "translate(" + (width - 15) + "," + (height/2) + ")" + "rotate(90)")
        .style("text-anchor", "middle")
        .style("font-size", font_size*1.2)
        .text("Time [Gyr]");

    // Lookback
    var temp = svg.append("text")
        .attr("id", "tl")
        .attr("transform", "translate(" + (width - 15) + "," + (height/2) + ")" + "rotate(90)")
        .style("text-anchor", "middle")
        .text("Lookback");

    // Shift up
    temp.attr("dx", -temp.node().getBBox().width/2 - 10)
        .attr("dy", temp.node().getBBox().height + 4)

    // Luminosity
    var temp = svg.append("text")
        .attr("id", "ta")
        .attr("transform", "translate(" + (width - 15) + "," + (height/2) + ")" + "rotate(90)")
        .style("text-anchor", "middle")
        .text("Universe");

    // Shift down
    temp.attr("dx", temp.node().getBBox().width/2 + 10)
        .attr("dy", temp.node().getBBox().height + 4)
}

function axisLabelTens(ax) {
    ax.selectAll(".tick text")
        .text(null)
        .filter(powerOfTen)
        .text(10)
        .append("tspan")
        .attr("dy", "-.7em")
        .text(function(d) {
            return sprintf('%+.0f', Math.round(Math.log(d) / Math.LN10));
        });
}

function powerOfTen(d) {
  return d / Math.pow(10, Math.ceil(Math.log(d) / Math.LN10 - 1e-12)) === 1;
}

// == Plot / Figure Updates == //

function parseRetVals(rv) {
    return String.fromCharCode.apply(null, rv);
}

function updateCrossHairs(retval) {
    console.log("plot.updateCrossHairs() : ", retval);
    var temp = parseFloat(retval['z']);
    xx = scale_z(temp) + margin.left;
    focus.select('#focusLine_z')
        .attr('x1', xx).attr('y1', 0)
        .attr('x2', xx).attr('y2', height);

    temp = parseFloat(parseRetVals(retval['dl']).replace(" Mpc", ""));
    yy = scale_d(temp) + margin.top;
    focus.select('#focusLine_dl')
        .attr('x1', 0).attr('y1', yy)
        .attr('x2', width).attr('y2', yy);

    temp = parseFloat(parseRetVals(retval['dc']).replace(" Mpc", ""));
    yy = scale_d(temp) + margin.top;
    focus.select('#focusLine_dc')
        .attr('x1', 0).attr('y1', yy)
        .attr('x2', width).attr('y2', yy);

    temp = parseFloat(parseRetVals(retval['tl']).replace(" Gyr", ""));
    yy = scale_t(temp) + margin.top;
    focus.select('#focusLine_tl')
        .attr('x1', 0).attr('y1', yy)
        .attr('x2', width).attr('y2', yy);

    temp = parseFloat(parseRetVals(retval['ta']).replace(" Gyr", ""));
    yy = scale_t(temp) + margin.top;
    focus.select('#focusLine_ta')
        .attr('x1', 0).attr('y1', yy)
        .attr('x2', width).attr('y2', yy);

}

function loadAndPlotCosmoLines() {
    d3.csv(DATA_FILE_PATH, function(error, _data) {
        if (error) throw error;
        data = _data;
        data_z = [];
        data_dl = [];
        data_len = data.length;
        for (var ii = 0; ii < data_len; ii++) {
            data_z.push(parseFloat(data[data_len - 1 - ii].z));
            data_dl.push(parseFloat(data[data_len - 1 - ii].dl));
        }

        var transLoc = "translate(" + (margin.left) + "," + (margin.top) + ")";
        data.splice(-1, 1);

        // format the data
        data.forEach(function(d, i) {
            d.z = parseFloat(d.z);
            d.dc = parseFloat(d.dc);
            d.dl = parseFloat(d.dl);
            d.tl = parseFloat(d.tl);
            d.ta = parseFloat(d.ta);
        });

        // ==  Comoving Distance d_C  ==
        // define the line
        line_dc = d3.line()
            .x(function(d) { return scale_z(d.z); })
            .y(function(d) { return scale_d(d.dc); });

        // Add the valueline path.
        path_dc = svg.select("#plots").append("path")
            .data([data])
            .attr("class", "line")
            .attr("id", "line_dc")
            .attr("d", line_dc)
            .attr("transform", transLoc);

        // ==  Luminosity Distance d_L  ==
        // define the line
        line_dl = d3.line()
            .x(function(d) { return scale_z(d.z); })
            .y(function(d) { return scale_d(d.dl); });

        // Add the valueline path.
        path_dl = svg.select("#plots").append("path")
            .data([data])
            .attr("class", "line")
            .attr("id", "line_dl")
            .attr("d", line_dl)
            .attr("transform", transLoc);

        // ==  Lookback Time t_l  ==
        // define the line
        line_tl = d3.line()
            .x(function(d) { return scale_z(d.z); })
            .y(function(d) { return scale_t(d.tl); });

        // Add the valueline path.
        path_tl = svg.select("#plots").append("path")
            .data([data])
            .attr("class", "line")
            .attr("id", "line_tl")
            .attr("d", line_tl)
            .attr("transform", transLoc);

        // ==  Universe-Age Time t_a  ==
        // define the line
        line_ta = d3.line()
            .x(function(d) { return scale_z(d.z); })
            .y(function(d) { return scale_t(d.ta); });

        // Add the valueline path.
        path_ta = svg.select("#plots").append("path")
            .data([data])
            .attr("class", "line")
            .attr("id", "line_ta")
            .attr("d", line_ta)
            .attr("transform", transLoc);

    });
}

// == Logistical Functions == //

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

function plot() {
    loadAndPlotCosmoLines();
    retval = true;
}

async function tryPlot(dataFileName) {
    do {
        if (sleep_count > SLEEP_COUNT_MAX) {
            console.log("Could not find data file after 20 iterations!");
            var temp = svg.append("text")
                .attr("class", "error")
                .attr("id", "loading_plot_data")
                .attr("transform",
                    "translate(" + (width/2) + ", " + (height/2) + ")")
                .style("text-anchor", "middle")
                .style("font-size", font_size*2);

            temp.append("tspan")
                .attr("x", 0)
                .attr("dy", "-40px")
                .text("Error loading plot data!");
            temp.append("tspan")
                .attr("x", 0)
                .attr("dy", "40px")
                .text("Try re-running?");
            break;
        }
        fs.exists(dataFileName, (exists) => {
            if (exists && retval == false) {
                plot();
            } else {
                console.log("File '", dataFileName, "' does not exist...");
                sleep_count++;
            }
        });
        if (retval == false) {
            await sleep(100);
        }
    } while (retval === false);
}

function loadDataFilePath() {

    var setsPath = "";
    var testPath = "";

    testPath = path.join(APP_DIR, 'external', 'settings.txt');
    console.log("Searching for settings at ", testPath);
    if (fs.existsSync(testPath)) {
        setsPath = testPath;
        console.log("Found settings!");
    } else {
        testPath = path.join(APP_DIR, '..', '..', 'external', 'settings.txt');
        console.log("Searching for settings at ", testPath);
        if (fs.existsSync(testPath)) {
            setsPath = testPath;
            console.log("Found settings!");
        } else {
            throw "Could not find settings file!";
        }
    }

    var data = fs.readFileSync(setsPath).toString().split("\n");
    var sets = {};
    for (var ii = 0; ii < data.length; ii++) {
        let clean = data[ii].trim();
        if (clean.length > 0) {
            clean = clean.split("=");
            if (clean.length === 2) {
                sets[clean[0].trim()] = clean[1].trim();
            }
        }
    }
    return APP_DIR + "/" + sets["FNAME_COSMO_DATA"];
}

/*    ========    RUN   =========    */

DATA_FILE_PATH = loadDataFilePath();
console.log("DATA_FILE_PATH = ", DATA_FILE_PATH);

initPlot();
tryPlot(DATA_FILE_PATH);


app/face/pyapi.py
"""API from `cosmo` electron application to `cosmopy` python package.
"""
import datetime
import os
import sys

import numpy as np
import zerorpc

import cosmopy

_DIR_APP = os.path.abspath(os.path.join(os.path.dirname(__file__), os.path.pardir))
_PATH_SETS = os.path.join(_DIR_APP, "settings.txt")
_MY_NAME = "pyapi.py"

_SETS_ILLEGAL_CHARS = ["\'", "\""]


#  ====  Load Settings and Log File  ====

def load_settings():
    if not os.path.exists(_PATH_SETS):
        print("{}:".format(_MY_NAME))
        raise IOError("Settings file missing '{}'".format(_PATH_SETS))

    contents = open(_PATH_SETS).read()
    check = True
    for ic in _SETS_ILLEGAL_CHARS:
        if ic in contents:
            check = False

    if not check:
        raise ValueError(
            "Error: Settings file '{}' should not contain quotation marks!".format(_PATH_SETS)
        )

    str_sets = [lin.strip() for lin in contents.split('\n')]
    sets = {}
    for lin in str_sets:
        if lin.startswith('#') or len(lin.strip()) == 0:
            continue
        kk, vv = [ll.strip() for ll in lin.split('=')]
        sets[kk] = vv

    return sets


def log(msg, stdout=False):
    dt = datetime.datetime.now().strftime('%Y/%m/%d-%H:%M:%S')
    str_out = "{} :: {} :: {}".format(dt, _MY_NAME, msg)
    if stdout:
        print(str_out)
    with open(PATH_LOG, 'a') as out:
        out.write(str_out + "\n")
        out.flush()
    return


try:
    sets = load_settings()
    FNAME_LOG = sets["FNAME_LOG"]
    FNAME_DATA = sets["FNAME_COSMO_DATA"]
    PATH_LOG = os.path.join(_DIR_APP, FNAME_LOG)
    PATH_DATA = os.path.join(_DIR_APP, FNAME_DATA)
    _data_dir = os.path.dirname(PATH_DATA)
    if not os.path.exists(_data_dir):
        os.makedirs(_data_dir)
    print("{} :: log path: '{}'".format(_MY_NAME, PATH_LOG))
except IOError:
    raise
except Exception:
    print("{}:".format(_MY_NAME))
    print("Failed to load settings!")
    raise

log("Started and initialized.", True)

log("Loading cosmology instance")
cosmo = cosmopy.get_cosmology()
log("\tcosmology instance loaded")


def main():
    log("pyapi.main()")
    save_cosmo_data()
    # sys.exit(232)

    addr = 'tcp://127.0.0.1:' + str(parse_port())
    s = zerorpc.Server(CalcApi())
    s.bind(addr)
    log('start running on {}'.format(addr))
    s.run()


def save_cosmo_data():
    log("pyapi.save_cosmo_data()")
    # Save the grid of cosmological values to a local file
    try:
        grid, names, units = cosmo.get_grid()
        np.savetxt(PATH_DATA, grid, delimiter=',', header=','.join(names), comments='')
    except:
        log("Failed to save cosmological data!")
        raise
    else:
        log("Saved to '{}' ({})".format(PATH_DATA, os.path.exists(PATH_DATA)), True)

    return


def parse_port():
    return 4242


class CalcApi:

    def __init__(self):
        log("CalcApi()")

    def calc(self, args):
        """based on the input text, return the int result"""
        beg = datetime.datetime.now()
        log("CalcApi.calc()")
        log("args = '{}'".format(args))
        msg = 'Calculation '

        try:
            name, _val = args
            log("name = '{}', val = '{}'".format(name, _val))
            rv = cosmopy.api(name, _val, cosmo=cosmo)
            msg += 'succeeded.'
        except Exception as e:
            log("\nException: '{}'\n".format(str(e)))
            msg += "failed.  Error: '{}'".format(str(e))

        msg += ' After {}'.format(datetime.datetime.now() - beg)
        log(msg)
        retval = {'dict': rv,
                  'msg': msg}
        return retval

    def echo(self, anything):
        """Test function for javascript interface.
        """
        log("CalcApi.echo()")
        return anything


if __name__ == '__main__':
    main()
    # CalcApi().calc(['z', '0.1'])


app/face/renderer.js

/* == Initialize Python Interface == */

const zerorpc = require("zerorpc");
let client = new zerorpc.Client(timeout=1000);
client.connect("tcp://127.0.0.1:4242");

client.invoke("echo", "server ready", (error, res) => {
    if(error || res !== 'server ready') {
      console.error("Failed to connect to python process in `renderer.js`!");
      console.error(error);
    } else {
      console.log("server is ready");
    }
});

// Load DOM objects from document
let redshift = document.querySelector('#redshiftInput');
let scale = document.querySelector('#scaleInput');
let comDist = document.querySelector('#comDistInput');
let lumDist = document.querySelector('#lumDistInput');
let lbkTime = document.querySelector('#lbkTimeInput');
let ageTime = document.querySelector('#ageTimeInput');

let dialog = document.querySelector('#dialog');


function valueInput(e) {
    console.log("valueInput");
    // If `Enter` is pressed
    if (e.keyCode == 13) {
        // Get the element calling the action (this should be an `input`)
        let srcElement = e.srcElement;

        // Load quantities from element
        let srcName = srcElement.name;
        let srcId = srcElement.id;
        let value = srcElement.value;
        console.log(srcName, srcId, value);
        console.log("client = ", client);

        calcAndUpdate(value, srcName);
        return false;
    }
}

function calcAndUpdate(value, src="z") {
    var retval;
    let args = [src, value];
    console.log("renderer.calcAndUpdate: invoking with args = ", args);
    client.invoke("calc", args, (error, res) => {
        retval = res['dict'];
        let message = res['msg'];
        let msg = 'Python: ' + message;
        console.log("srcName = ", src);
        if (error) {
            console.log("Error!");
            console.error(error);
        } else {
            updateValues(retval, msg);
            updateCrossHairs(retval);
        }
    });
}

function updateValues(retval, msg) {
    console.log("renderer.updateValues()");
    redshift.value = retval['z'];
    scale.value = retval['a'];
    comDist.value = retval['dc'];
    lumDist.value = retval['dl'];
    lbkTime.value = retval['tl'];
    ageTime.value = retval['ta'];
    dialog.textContent = msg;
}

// Bind the function to DOM input forms
document.getElementById('redshiftForm').onkeypress = valueInput;
document.getElementById('scaleForm').onkeypress = valueInput;
document.getElementById('comDistForm').onkeypress = valueInput;
document.getElementById('lumDistForm').onkeypress = valueInput;
document.getElementById('lbkTimeForm').onkeypress = valueInput;
document.getElementById('ageTimeForm').onkeypress = valueInput;


app/main.js
const {app, BrowserWindow, shell, dialog} = require('electron');
const path = require('path');
const url = require('url');
const fs = require('fs');

require('electron-debug')({showDevTools: true, enabled: true});

var win;

var pyProc = null;
var pyPort = null;
var pyPath = null;


const fixPath = require('fix-path');
console.log("path before = ", process.env.PATH);
fixPath();
console.log("path after = ", process.env.PATH);


function initPython() {
    const selectPort = () => {
        pyPort = 4242;
        return pyPort;
    }

    const pathPyAPI = () => {
        pyPath = path.join(__dirname, 'external', 'pyapi.py');
        console.log("Trying to find python at ", pyPath);
        if (fs.existsSync(pyPath)) {
            console.log("\tLocated python!");
            return pyPath;
        }

        pyPath = path.join(__dirname, '..', '..', 'external', 'pyapi.py');
        console.log("Trying to find python at ", pyPath);
        if (fs.existsSync(pyPath)) {
            console.log("\tLocated python!");
            return pyPath;
        }

        console.log("Failed to find python!");
        throw "Failed to find python!";
    }

    const createPyProc = () => {
        let port = '' + selectPort();
        let pyPath = '' + pathPyAPI();
        console.log("pyPath: '", pyPath, "'", "on port", port)
        pyProc = require('child_process').spawn('python', [pyPath, port]);
        // pyProc = require('child_process').spawnSync('python', [pyPath, port]);
        if (pyProc != null) {
            console.log('child process success');
        }

    }

    const exitPyProc = () => {
        pyProc.kill();
        pyProc = null;
        pyPort = null;
    }

    app.on('ready', createPyProc);
    app.on('will-quit', exitPyProc);
}

function createWindow () {
    // Create the browser window.
    win = new BrowserWindow({
        width: 950,
        height: 500,
        x: 20,
        y: 40,
        resizeable: false,
        show: false,   // dont show until ready (below)
    });

    win.setResizable(false);
    // win.webContents.openDevTools();

    // and load the index.html of the app.
    win.loadURL(url.format({
        pathname: path.join(__dirname, 'index.html'),
        protocol: 'file:',
        slashes: true
    }))

    win.once('ready-to-show', () => {
        win.show()
    })

    // Emitted when the window is closed.
    win.on('closed', () => {
        win = null
    })

    // Load all external links in default browser
    win.webContents.on('will-navigate', function(e, reqUrl) {
        let getHost = url=>require('url').parse(url).host;
        let reqHost = getHost(reqUrl);
        let isExternal = reqHost && reqHost != getHost(win.webContents.getURL());
        console.log("host = ", getHost, reqHost);
        console.log(isExternal);
        if (isExternal) {
            e.preventDefault();
            shell.openExternal(reqUrl);
        }
    });
}

function initApp() {
    app.on('ready', createWindow);

    // Quit when all windows are closed.
    app.on('window-all-closed', () => {
        app.quit()
    });
}

// dialog.showErrorBox("hello", "content");
// exit(2);

/* ==  Create Python Process  == */
console.log("Calling `initPython()`");
initPython();

/* ==  Create App Window  == */
console.log("Calling `initApp()`");
initApp();
